### PR TITLE
Mesh seeding (surface)

### PR DESCRIPTION
**Pull request by StongeEtienne**
Utility functions, to seed from a mesh, using triangles and vertices,

Similar to #1635 
@kristianeschenburg , @Garyfallidis

### CODE REVIEW

**Comment by StongeEtienne** [comment]
Since I plan to include multiple utility functions for Surface (mesh) integration:
- surface/streamlines intersections
- mask generation (from annot / labels)
  etc ...

Maybe surfaces tools should be in a separate file or folder ?

**Comment by arokem** [comment]
We can move this later (but before 1.0). Could you please write some tests for this?

**Comment by gabknight** [comment]
@StongeEtienne how should we move forward with this PR and PR #1627. Do you think returning both the normal direction to the surface and the seed location is useful? So we can provide both the seed location and the initial direction to the tracking from any surface?

**Comment by StongeEtienne** [comment]
@gabknight Sorry I missed the meeting (I have a cold),
I currently use both #1627 and #1635 , for me it is useful, I still feel PFT sometime backtrack to the initial point and doesn't use the given direction. My SET initial direction is not perfect either and maybe I should find a more "aligned" direction. 

We could plan a meeting next week if needed.

**Comment by skoudoro** [comment]
Hi @StongeEtienne,

What is the status of this PR? Did you have the meeting with @gabknight? 
Any chance to have this ready for the release at the end of this month?

**Comment by gabknight** [comment]
Let's move this to the next release.

**Comment by StongeEtienne** [comment]
@skoudoro @arokem 
After discussion with @gabknight  I will refactor this PR,
and afterwards, once this Surface seeding is in DIpy,
Gabriel will work on the #1627, to start tracking from the initial dirirection 
(generated from the surface).

**Comment by arokem** [comment]
Excellent! Let us know when you would like us to take another look.

**Comment by StongeEtienne** [comment]
@skoudoro @arokem  Would it be easier to restart a new branch from the current Dipy Master ?
and gradually adding change in commit, I might also check to add some of the proposed #1635 . 
Thanks,

**Comment by skoudoro** [comment]
Rebasing should not be a problem but if you feel more comfortable with a new branch and many small PRs, fine with me too @StongeEtienne.

**Comment by StongeEtienne** [comment]
@skoudoro 
Hi serge, I updated the code (move to a new file) and added a doc/exemple and  testing function!
The build failed, should I update the main dipy version or ?

**Comment by skoudoro** [comment]
As you can see [here](https://dev.azure.com/dipy/dipy/_build/results?buildId=935&view=logs&j=09f0a5bc-9cbc-582e-eae5-2a9f4ac9c7a3&t=cce7cd89-7947-524c-1d49-2d925884a92a&l=2721)  you have to update one test related to the fetcher.

Otherwise, is it ready for review @StongeEtienne?

Thank you for updating your PR!

**Comment by skoudoro** [comment]
Hi @StongeEtienne,

Could you rebase? It seems that the build history has been deleted so I cannot point anymore the failing test.
It was a test related to `get_fnames` but I can not remember anymore. 

Thank you

File: dipy/tracking/tests/test_mesh.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        # if the current weighted vts is in the current triangle add 1
+        for j in range(len(cube_tri)):
+            if i in cube_tri[j]:
+                tri_w_manual[j] += 1.0
+
+        npt.assert_array_equal(tri_w_func, tri_w_manual)
+
+
+if __name__ == "__main__":
+    npt.run_module_suite()
```
Comment (skoudoro) [inline dipy/tracking/tests/test_mesh.py:85]: This part is not needed anymore. you can remove it

**Review by skoudoro — COMMENTED** [review]
Hi @StongeEtienne,

Below are the errors. They look easy to fix. You just need to update your doctests and 1 test.

```python
2021-02-23T20:00:54.3970051Z _________ [doctest] dipy.tracking.mesh.random_coordinates_from_surface _________
2021-02-23T20:00:54.3970756Z 029     trilin_coord: [s,3] array
2021-02-23T20:00:54.3971411Z 030         Randomly chosen trilinear_coordinates
2021-02-23T20:00:54.3972062Z 031 
2021-02-23T20:00:54.3972477Z 032     See Also
2021-02-23T20:00:54.3973181Z 033     --------
2021-02-23T20:00:54.3973765Z 034     seeds_from_surface_coordinates, random_seeds_from_mask
2021-02-23T20:00:54.3974323Z 035 
2021-02-23T20:00:54.3975087Z 036     Examples
2021-02-23T20:00:54.3976149Z 037     --------
2021-02-23T20:00:54.3976718Z 038     >>> mask = np.ones([len(triangles)])
2021-02-23T20:00:54.3977587Z UNEXPECTED EXCEPTION: NameError("name 'triangles' is not defined")
2021-02-23T20:00:54.3978430Z Traceback (most recent call last):
2021-02-23T20:00:54.3979547Z   File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/doctest.py", line 1337, in __run
2021-02-23T20:00:54.3980352Z     compileflags, 1), test.globs)
2021-02-23T20:00:54.3980934Z   File "<doctest dipy.tracking.mesh.random_coordinates_from_surface[0]>", line 1, in <module>
2021-02-23T20:00:54.3981732Z NameError: name 'triangles' is not defined
2021-02-23T20:00:54.3982687Z /home/vsts/work/1/s/venv/lib/python3.7/site-packages/dipy/tracking/mesh.py:38: UnexpectedException
2021-02-23T20:00:54.3984527Z _________ [doctest] dipy.tracking.mesh.seeds_from_surface_coordinates __________
2021-02-23T20:00:54.3985262Z 095     pts : [s, ...] array
2021-02-23T20:00:54.3985718Z 096         Interpolated values of vertices with triangles_idx and trilinear_coord
2021-02-23T20:00:54.3986144Z 097 
2021-02-23T20:00:54.3986455Z 098     See Also
2021-02-23T20:00:54.3987072Z 099     --------
2021-02-23T20:00:54.3987481Z 100     random_coordinates_from_surface
2021-02-23T20:00:54.3987841Z 101 
2021-02-23T20:00:54.3988145Z 102     Examples
2021-02-23T20:00:54.3988647Z 103     --------
2021-02-23T20:00:54.3989085Z 104     >>> seeds_pts = seeds_from_surface_coordinates(triangles, vts,
2021-02-23T20:00:54.3989782Z UNEXPECTED EXCEPTION: NameError("name 'triangles' is not defined")
2021-02-23T20:00:54.3990258Z Traceback (most recent call last):
2021-02-23T20:00:54.3990745Z   File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/doctest.py", line 1337, in __run
2021-02-23T20:00:54.3991190Z     compileflags, 1), test.globs)
2021-02-23T20:00:54.3991827Z   File "<doctest dipy.tracking.mesh.seeds_from_surface_coordinates[0]>", line 1, in <module>
2021-02-23T20:00:54.3992511Z NameError: name 'triangles' is not defined
2021-02-23T20:00:54.3993701Z /home/vsts/work/1/s/venv/lib/python3.7/site-packages/dipy/tracking/mesh.py:104: UnexpectedException
2021-02-23T20:00:54.3994672Z _______________________ test_io_fetch_fetcher_datanames ________________________
2021-02-23T20:00:54.3995026Z 
2021-02-23T20:00:54.3995412Z     def test_io_fetch_fetcher_datanames():
2021-02-23T20:00:54.3995899Z         available_data = FetchFlow.get_fetcher_datanames()
2021-02-23T20:00:54.3996298Z     
2021-02-23T20:00:54.3996968Z         dataset_names = ['bundle_atlas_hcp842', 'bundle_fa_hcp',
2021-02-23T20:00:54.3997754Z                          'bundles_2_subjects', 'cenir_multib', 'cfin_multib',
2021-02-23T20:00:54.3998849Z                          'file_formats', 'gold_standard_io', 'isbi2013_2shell',
2021-02-23T20:00:54.3999741Z                          'ivim', 'mni_template', 'qtdMRI_test_retest_2subjects',
2021-02-23T20:00:54.4000571Z                          'scil_b0', 'sherbrooke_3shell', 'stanford_hardi',
2021-02-23T20:00:54.4001552Z                          'stanford_labels', 'stanford_pve_maps', 'stanford_t1',
2021-02-23T20:00:54.4002395Z                          'syn_data', 'taiwan_ntu_dsi', 'target_tractogram_hcp',
2021-02-23T20:00:54.4003496Z                          'tissue_data']
2021-02-23T20:00:54.4003974Z     
2021-02-23T20:00:54.4004415Z         num_expected_fetch_methods = len(dataset_names)
2021-02-23T20:00:54.4004979Z >       npt.assert_equal(len(available_data), num_expected_fetch_methods)
2021-02-23T20:00:54.4005482Z E       AssertionError: 
2021-02-23T20:00:54.4005949Z E       Items are not equal:
2021-02-23T20:00:54.4006521Z E        ACTUAL: 22
2021-02-23T20:00:54.4006918Z E        DESIRED: 21
```

File: dipy/tracking/tests/test_mesh.py (hunk 1 above)
Comment (StongeEtienne) [reply dipy/tracking/tests/test_mesh.py:85]: both lines or just the  "npt.run_module_suite()" ?

Comment (skoudoro) [reply dipy/tracking/tests/test_mesh.py:85]: both line

File: dipy/tracking/mesh.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    trilin_coord: [s,3] array
+        Randomly chosen trilinear_coordinates
+
+    See Also
+    --------
+    seeds_from_surface_coordinates, random_seeds_from_mask
+
+    Examples
+    --------
+    >>> mesh_polydata = load_polydata('mesh.vtk') # fury.io
```
Comment (skoudoro) [inline dipy/tracking/mesh.py:38]: Doctests will not recognize these functions. you need to add the `import` above as usual.

```python
>>> from fury import io, utils
>>> mesh_polydata = io.load_polydata('mesh.vtk')
etc...
```

File: dipy/tracking/mesh.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    pts : [s, ...] array
+        Interpolated values of vertices with triangles_idx and trilinear_coord
+
+    See Also
+    --------
+    random_coordinates_from_surface
+
+    Examples
+    --------
+    >>> mesh_polydata = load_polydata('mesh.vtk') # fury.io
```
Comment (skoudoro) [inline dipy/tracking/mesh.py:107]: Same comment as above. Doctests will not recognize these functions. you need to add the `import` above as usual.

```python
>>> from fury import io, utils   # skip if not HAVE_FURY
>>> mesh_polydata = io.load_polydata('mesh.vtk')  # skip if not HAVE_FURY
etc...
```

Also, you might have to use the skip directive in both case `# skip if not HAVE_FURY`  
HAVE_FURY, should be defined above with the main  import

**Review by skoudoro — COMMENTED** [review]
Thanks for the update @StongeEtienne.

See below why it fails.

File: dipy/tracking/mesh.py (hunk 3 above)
Comment (StongeEtienne) [reply dipy/tracking/mesh.py]: I tried but doest seem to work

Comment (skoudoro) [reply dipy/tracking/mesh.py]: Strange, I need to test locally.

File: dipy/tracking/mesh.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
+
+import numpy as np
+
+from dipy.utils.optpkg import optional_package
+
```
Comment (skoudoro) [inline dipy/tracking/mesh.py:5]: Also, import `from dipy.testing.decorators import doctest_skip_parser`

File: dipy/tracking/mesh.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+_, has_fury, _ = optional_package('fury')
+
+def random_coordinates_from_surface(nb_triangles, nb_seed, triangles_mask=None,
```
Comment (skoudoro) [inline dipy/tracking/mesh.py:8]: After the import above, add the following decorator above the function: `@doctest_skip_parser`

File: dipy/tracking/mesh.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    # http://mathworld.wolfram.com/TrianglePointPicking.html
+    trilin_coord = np.random.rand(nb_seed, 3)
+    is_upper = (trilin_coord[:, 1:].sum(axis=-1) > 1.0)
+    trilin_coord[is_upper] = 1.0 - trilin_coord[is_upper]
+    trilin_coord[:, 0] = 1.0 - (trilin_coord[:, 1] + trilin_coord[:, 2])
+
+    return triangles_idx, trilin_coord
+
+
+def seeds_from_surface_coordinates(triangles, vts_values,
```
Comment (skoudoro) [inline dipy/tracking/mesh.py:84]: After the import above, add the following decorator above the function: `@doctest_skip_parser`

File: dipy/tracking/mesh.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        Randomly chosen trilinear_coordinates
+
+    See Also
+    --------
+    seeds_from_surface_coordinates, random_seeds_from_mask
+
+    Examples
+    --------
+    >>> from fury import io, utils   # skip if not has_fury
+    >>> mesh_polydata = io.load_polydata('mesh.vtk') # skip if not has_fury
```
Comment (skoudoro) [inline dipy/tracking/mesh.py:42]: what is `mesh.vtk` ? the cannot reach this data

File: dipy/tracking/mesh.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        Interpolated values of vertices with triangles_idx and trilinear_coord
+
+    See Also
+    --------
+    random_coordinates_from_surface
+
+    Examples
+    --------
+    >>> from fury import io, utils   # skip if not has_fury
+    >>> mesh_polydata = io.load_polydata('mesh.vtk') # skip if not has_fury
```
Comment (skoudoro) [inline dipy/tracking/mesh.py:112]: what is `mesh.vtk` ? the cannot reach this data

File: dipy/tracking/mesh.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    >>> mask = np.ones([len(triangles)])
+    >>> mask[0] = 0
+    >>> normal = np.cross(vts[triangles[:, 1]] - vts[triangles[:, 0]],
+    ... vts[triangles[:, 2]] - vts[triangles[:, 0]])
+    >>> triangles_area = 0.5*np.linalg.norm(normal,axis=-1)
+    >>> nb_seeds = 100
+    >>> idx, coord = random_coordinates_from_surface(len(triangles), nb_seeds,
+    ... mask, triangles_weight=triangles_area)
+    >>> seeds_pts = seeds_from_surface_coordinates(triangles, vts, idx,
+    ... coord)
```
Comment (skoudoro) [inline dipy/tracking/mesh.py:54]: replace the whole test by the following: 

```python
    >>> def demo_random_coords():
    ...     from fury import io, utils
    ...     mesh_polydata = io.load_polydata('mesh.vtk')
    ...     vts = utils.get_polydata_vertices(mesh_polydata)
    ...     triangles = utils.get_polydata_triangles(mesh_polydata)
    ...     mask = np.ones([len(triangles)])
    ...     mask[0] = 0
    ...     normal = np.cross(vts[triangles[:, 1]] - vts[triangles[:, 0]],
    ...         vts[triangles[:, 2]] - vts[triangles[:, 0]])
    ...     triangles_area = 0.5*np.linalg.norm(normal,axis=-1)
    ...     nb_seeds = 100
    ...     idx, coord = random_coordinates_from_surface(len(triangles),
    ...         nb_seeds, mask, triangles_weight=triangles_area)
    ...     seeds_pts = seeds_from_surface_coordinates(triangles, vts, idx,
    ...         coord)
    ...
    >>> demo_random_coords()   # skip if not has_fury
```

File: dipy/tracking/mesh.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    >>> from fury import io, utils   # skip if not has_fury
+    >>> mesh_polydata = io.load_polydata('mesh.vtk') # skip if not has_fury
+    >>> vts = utils.get_polydata_vertices(mesh_polydata) # skip if not has_fury
+    >>> triangles = utils.get_polydata_triangles(mesh_polydata) # skip if not has_fury
+    >>> seeds_pts = seeds_from_surface_coordinates(triangles, vts,
+    ... triangles_idx, triangles_coord)
+    >>> vts_normal = utils.normals_from_v_f(vts, triangles) # skip if not has_fury
+    >>> seeds_normal = seeds_from_surface_coordinates(triangles, vts_normal,
+    ... triangles_idx, triangles_coord)
+    >>> seeds_normal /= np.linalg.norm(seeds_normal, axis=-1, keepdims=True)
```
Comment (skoudoro) [inline dipy/tracking/mesh.py:120]: replace the whole test by the following:
```python
    >>> def demo_seed():
    ...     from fury import io, utils
    ...     mesh_polydata = io.load_polydata('mesh.vtk')
    ...     vts = utils.get_polydata_vertices(mesh_polydata)
    ...     triangles = utils.get_polydata_triangles(mesh_polydata)
    ...     seeds_pts = seeds_from_surface_coordinates(triangles, vts,
    ...         triangles_idx, triangles_coord)
    ...     vts_normal = utils.normals_from_v_f(vts, triangles)
    ...     seeds_normal = seeds_from_surface_coordinates(triangles,
    ...         vts_normal, triangles_idx, triangles_coord)
    ...     seeds_normal /= np.linalg.norm(seeds_normal, axis=-1,
    ...         keepdims=True)
    ...
    >>> demo_seed()  # skip if not has_fury
```

**Review by skoudoro — COMMENTED** [review]
Hi @StongeEtienne,

Sorry to get back late to you. See below for all the updates that you need to make it pass.
Also, your `mesh.vtk` file should be defined or downloaded somewhere. Consider doctest as a real test like your file `test_mesh.py`. So it follows the same rules.

Let me know if you have any question

**Comment by StongeEtienne** [comment]
@skoudoro 
since there is an example, should we simply remove the example from the docstring and point to the mesh seed example ?

**Comment by skoudoro** [comment]
it can be a solution

**Comment by skoudoro** [comment]
This PR seems ready to go! 

Does someone want to have a last look at this PR?

**Comment by skoudoro** [comment]
ok, no comment, Thanks @StongeEtienne! merging

**Pull request merged**